src/voice/voice.cpp

```diff
@@ -253,28 +253,81 @@ template <bool OS> bool Voice::processWithOS()
     modMatrix.process();
 
     auto fpitch = calculateVoicePitch();
-    calculateGeneratorRatio(fpitch);
-    if (useOversampling)
-        GD.ratio = GD.ratio >> 1;
-    fpitch -= 69;
 
-    // TODO : Start and End Points
-    if (sampleIndex >= 0)
-    {
-        auto &s = zone->samplePointers[sampleIndex];
-        assert(s);
+    auto [firstIndex, lastIndex] = sampleIndexRange();
+    for (auto i = firstIndex; i < lastIndex; ++i)
+        if (zone->samplePointers[i])
+            calculateGeneratorRatio(fpitch, i, i - firstIndex);
 
-        GD.sampleStart = 0;
-        GD.sampleStop = s->sample_length;
+    if (useOversampling)
+        for (auto i = 0; i < numGeneratorsActive; ++i)
+            GD[i].ratio = GD[i].ratio >> 1;
+    fpitch -= 69;
```

Comment:
Correction factor or debug offset? 🤣

Reply:
It’s a 440 vs an 8.12 convention (which is a mess in surge because the function names are the same). But the same offset was in the single generator case outside the loop above